src/pact/v3/generate/generator.py

```diff
@@ -0,0 +1,161 @@
+"""
+Implementations of generators for the V3 and V4 specifications.
+"""
+
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from itertools import chain
+from typing import TYPE_CHECKING, Any, Mapping
+
+if TYPE_CHECKING:
+    from pact.v3.types import GeneratorType
+
+
+class Generator(ABC):
+    """
+    Abstract generator.
+
+    In Pact, a generator is used by Pact to generate data on-the-fly during the
+    contract verification process. Generators are used in combination with
+    matchers to provide more flexible matching of data.
+
+    This class is abstract and should not be used directly. Instead, use one of
+    the concrete generator classes. Alternatively, you can create your own
+    generator by subclassing this class
+
+    The matcher provides methods to convert into an integration JSON object and
+    a matching rule. These methods are used internally by the Pact library when
+    communicating with the FFI and generating the Pact file.
+    """
+
+    @abstractmethod
+    def to_integration_json(self) -> dict[str, Any]:
+        """
+        Convert the matcher to an integration JSON object.
+
+        This method is used internally to convert the matcher to a JSON object
+        which can be embedded directly in a number of places in the Pact FFI.
+
+        For more information about this format, see the docs:
+
+        > https://docs.pact.io/implementation_guides/rust/pact_ffi/integrationjson
+
+        Returns:
+            The matcher as an integration JSON object.
+        """
+
+    @abstractmethod
+    def to_generator_json(self) -> dict[str, Any]:
+        """
+        Convert the generator to a generator JSON object.
+
+        This method is used internally to convert the generator to a JSON object
+        which can be embedded directly in a number of places in the Pact FFI.
+
+        For more information about this format, see the docs:
+
+        > https://github.com/pact-foundation/pact-specification/tree/version-4
+
+        and
+
+        > https://github.com/pact-foundation/pact-specification/tree/version-2?tab=readme-ov-file#matchers
+
+        Returns:
+            The generator as a generator JSON object.
+        """
+
+
+class GenericGenerator(Generator):
+    """
+    Generic generator.
+
+    A generic generator, with the ability to specify the generator type and
+    additional configuration elements.
+    """
+
+    def __init__(
+        self,
+        type: GeneratorType,  # noqa: A002
+        /,
+        extra_fields: Mapping[str, Any] | None = None,
+        integration_fields: Mapping[str, Any] | None = None,
+        generator_json_fields: Mapping[str, Any] | None = None,
```

Comment:
Is it possible to provide examples of how these two might be used?

Reply:
Yeah, good call.

The reasoning initially is that the way the generators/matchers are serialised into the 'integration JSON' format is slightly different to the way they are serialised when passed directly to the `matching_rules` or `generators` commands. E.g., in one, you have to use `"pact:matcher:type": "..."` whereas the other you don't.

As far as I know, all the other fields are the same, but if there were others that had to be differentiated, then this is there.

Thinking back on it though, I'm concerned that this might be a case of premature generalisation.